packages/snjs/lib/services/api/api_service.ts

```diff
@@ -207,31 +212,80 @@ export class SNApiService extends PureService {
     }
   }
 
-  /**
-   * @param mfaKeyPath  The params path the server expects for authentication against
-   *                    a particular mfa challenge. A value of foo would mean the server
-   *                    would receive parameters as params['foo'] with value equal to mfaCode.
-   * @param mfaCode     The mfa challenge response value.
-   */
-  getAccountKeyParams(
+  private async getAuthMethods(
+    email: string,
+  ): Promise<HttpResponse<GetAuthMethodsResponse>> {
+    const params = this.params({ email });
+
+    return this.request({
+      verb: HttpVerb.Get,
+      url: joinPaths(this.host, Paths.v0.authMethods),
+      fallbackErrorMessage: messages.UNKNOWN_ERROR,
+      params,
+    }) as Promise<HttpResponse<GetAuthMethodsResponse>>
+  }
+
+  private async requestWithMfa(
+    httpRequest: HttpRequestWithFallbackError & {
+      identifier: string,
+    },
+  ): Promise<HttpResponse> {
+    const {identifier, ...restRequest} = httpRequest
+    for (;;) {
+      const authMethodsResponse = await this.getAuthMethods(identifier)
+      const mfaResult = await this.mfaService.handleMfa(authMethodsResponse)
+
+      switch (mfaResult.status) {
+        case MfaStatus.ErrorMissingCode: {
+          // User dismissed window without input
+          return this.createErrorResponse(
+            messages.SignInStrings.SignInCanceledMissingMfa,
+            StatusCode.CanceledMfa,
+          )
+        }
+        case MfaStatus.ErrorResponse: {
+          return this.createErrorResponse(
+            messages.SignInStrings.ErrorFetchingAuthMethods,
+            StatusCode.UnknownError,
+          )
+        }
+        default: {
+          const mfaParams = this.mfaService.getMfaQueryParams(mfaResult)
+      
+          const response = await this.request({
+            ...restRequest,
+            params: {
+              ...restRequest.params,
+              ...mfaParams,
+            }
+          })
+        
+          if (response.error?.tag === ErrorTag.InvalidMfaCode) {
+            await this.alertService.alert(messages.SignInStrings.IncorrectMfa)
```

Comment:
Not sure if this needs to be awaited. I thinking awaiting would wait on the user to dismiss the alert before continuing code execution (unless that's what you want?)

Reply:
Same thing was happening previously:

```js
      if (signInResponse.error.payload?.mfa_key) {
        if (mfaCode) {
          await this.alertService.alert(SignInStrings.IncorrectMfa);
        }
```

`if (mfaCode)` would be true only on retry (so not the first time there was an error response) -- so the logic here is the same, except now it's more explicit.

I assume this is the intended behavior. If not -- let me know and I can change it to not await.